src/socket.c

```diff
@@ -619,17 +649,60 @@ mrb_tcpsocket_allocate(mrb_state *mrb, mrb_value klass)
   return mrb_obj_value((struct RObject*)mrb_obj_alloc(mrb, ttype, c));
 }
 
+/* Windows overrides for IO methods on BasicSocket objects.
+ * This is because sockets on Windows are not the same as file
+ * descriptors, and thus functions which operate on file descriptors
+ * will break on socket descriptors.
+ */
+#ifdef _WIN32
+static mrb_value
+mrb_win32_basicsocket_close(mrb_state *mrb, mrb_value self)
+{
+  if (closesocket(socket_fd(mrb, self)) != NO_ERROR)
+    mrb_sys_fail(mrb, "closesocket");
+  return mrb_nil_value();
+}
+
+static mrb_value
+mrb_win32_basicsocket_write(mrb_state *mrb, mrb_value self)
+{
+  int n;
+  SOCKET sd;
+  mrb_value str;
+
+  sd = socket_fd(mrb, self);
+  mrb_get_args(mrb, "S", &str);
+  n = send(sd, RSTRING_PTR(str), RSTRING_LEN(str), 0);
+  if (n == SOCKET_ERROR)
+    mrb_sys_fail(mrb, "write");
```

Comment:
Yes. On Windows, socket descriptors are not file descriptors. If you try to call write() or fwrite() on it, it'll cause issues. Nonetheless, in normal Ruby, all of the traditional socket classes have a working write() function. It stands to reason that the function should be overridden for the BasicSocket class so that it can use send() to nonetheless work in the same way.